### PR TITLE
Initial integration of libdispatch into build-script

### PR DESCRIPTION
**Pull request by dgrove-oss**
Extend build-script, build-script-impl, and update-checkout
to include libdispatch.  For now, libdispatch is not
built by default (user must enable via command line
argument).

Integration of testing is functional, but should be improved
in a later pull request.  The basic autotools based test
harness does not give the nice high-level progress output
as the rest of the test suite.

A related pull request to libdispatch (#34) has some fixes
to the autotools build that are needed to enable the test
target to succeed when run in an external directory.

### CODE REVIEW

**Comment by dgrove-oss** [comment]
Taking the approach as discussed with @parkera on the swift-corelibs-dev list yesdterday of just invoking libdisaptch's autotools build machinery from these scripts.

File: utils/build-script-impl [hunk 1]
```diff
@@ -1284,6 +1300,9 @@ function cmake_config_opt() {
             foundation)
                 echo "--config ${FOUNDATION_BUILD_TYPE}"
                 ;;            
+            libdispatch)
```
Comment (gribozavr) [inline utils/build-script-impl]: Does libdispatch use CMake?  If not, why is this change needed?

File: utils/build-script-impl [hunk 2]
```diff
@@ -2013,6 +2054,21 @@ for deployment_target in "${STDLIB_DEPLOYMENT_TARGETS[@]}"; do
...
                 continue
                 ;;
+            libdispatch)
+                if [[ "${SKIP_TEST_LIBDISPATCH}" ]]; then
+                    continue
+                fi
+
+                LIBDISPATCH_BUILD_DIR=$(build_directory ${deployment_target} ${product})
+
+				set -x
```
Comment (gribozavr) [inline utils/build-script-impl]: Indentation is off here.

File: utils/build-script-impl [hunk 3]
```diff
@@ -2013,6 +2054,21 @@ for deployment_target in "${STDLIB_DEPLOYMENT_TARGETS[@]}"; do
                 echo "--- Finished tests for ${product} ---"
                 continue
                 ;;
+            libdispatch)
+                if [[ "${SKIP_TEST_LIBDISPATCH}" ]]; then
+                    continue
+                fi
+
+                LIBDISPATCH_BUILD_DIR=$(build_directory ${deployment_target} ${product})
```
Comment (gribozavr) [inline utils/build-script-impl]: Would you mind printing a "Running tests" banner like other cases do?  This makes CI output much better.

File: utils/build-script-impl [hunk 4]
```diff
@@ -1826,6 +1845,28 @@ for deployment_target in "${NATIVE_TOOLS_DEPLOYMENT_TARGETS[@]}" "${CROSS_TOOLS_
                 # Foundation builds itself and doesn't use cmake
                 continue
                 ;;
+			libdispatch)
+                LIBDISPATCH_BUILD_DIR=$(build_directory ${deployment_target} ${product})
+
+				set -x
```
Comment (gribozavr) [inline utils/build-script-impl]: Indentation seems off.

File: utils/build-script-impl [hunk 5]
```diff
@@ -1826,6 +1845,28 @@ for deployment_target in "${NATIVE_TOOLS_DEPLOYMENT_TARGETS[@]}" "${CROSS_TOOLS_
                 # Foundation builds itself and doesn't use cmake
                 continue
                 ;;
+			libdispatch)
```
Comment (gribozavr) [inline utils/build-script-impl]: Indentation seems off.

File: utils/build-script-impl [hunk 6]
```diff
@@ -1826,6 +1845,28 @@ for deployment_target in "${NATIVE_TOOLS_DEPLOYMENT_TARGETS[@]}" "${CROSS_TOOLS_
...
+
+				set -x
+				if [[ ! -f "${LIBDISPATCH_BUILD_DIR}"/config.status ]]; then
+					# First time building; need to run autotools and configure
+					mkdir -p "${LIBDISPATCH_BUILD_DIR}"
+					pushd "${LIBDISPATCH_SOURCE_DIR}"
+					autoreconf -fvi
+					popd
+					pushd "${LIBDISPATCH_BUILD_DIR}"
+					"${LIBDISPATCH_SOURCE_DIR}"/configure --prefix="${INSTALL_DESTDIR}"/"${INSTALL_PREFIX}"
```
Comment (gribozavr) [inline utils/build-script-impl]: Do we need to pass the build type (debug/release/assertions) here?

File: utils/build-script-impl [hunk 7]
```diff
@@ -1184,6 +1197,9 @@ function build_directory_bin() {
             foundation)
                 echo "${root}/${FOUNDATION_BUILD_TYPE}/bin"
                 ;;
+            libdispatch)
+                echo "${root}/${LIBDISPATCH_BUILD_TYPE}/bin"
```
Comment (gribozavr) [inline utils/build-script-impl]: libdispatch does not use CMake with Xcode generator.  This should return `"${root}/bin"`, if that's the correct binary directory for libdispatch.

**Comment by gribozavr** [comment]
Some comments above, but looks reasonable in general.

File: utils/build-script-impl (hunk 6 above)
Comment (dgrove-oss) [inline utils/build-script-impl]: I put in some of the plumbing at higher levels of the scripts to support multiple build types following what is done for foundation, but right now it isn't being used by the autotools driven build.  I think it makes sense to eventually have multiple build types, but I'm not 100% sure exactly how it will be passed down, so leaving for a latter commit.

**Comment by dgrove-oss** [comment]
Thanks for the quick review.  Sorry about the indentation;  I was using my libdispatch emacs mode (force tabs; tabs are 4 spaces).  Need a different mode for foundation...
Pushed a follow up commit to hopefully address all your comments.

**Comment by gribozavr** [comment]
Thanks.  One more place with tabs.

My concern about the build mode is that we shouldn't accidentally ship a debug library, and IIRC Autoconf builds in debug by default.  We can fix that in a later commit though.

**Comment by dgrove-oss** [comment]
Thanks.  I'll look on the dispatch side and see if it needs tweaking to support debug/release build modes in the autoconf driven build.

Do you need me to squash & open a new pull request (since this one has gotten a bit messy with the whitespace fixups)?

**Comment by gribozavr** [comment]
@dgrove-oss Please squash in-place.

**Comment by dgrove-oss** [comment]
squashed in place.  Thanks.

**Comment by gribozavr** [comment]
Passes tests on Ubuntu 14.04.  Merging.
